src/estimators/meta_estimators.py

```diff
@@ -0,0 +1,171 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Implementations of estimators which count Bloom Filter indices."""
+import copy
+
+import numpy as np
+from wfa_cardinality_estimation_evaluation_framework.estimators.base import EstimatorBase
+import wfa_cardinality_estimation_evaluation_framework.estimators.bloom_filters as bf
+import wfa_cardinality_estimation_evaluation_framework.estimators.vector_of_counts as voc
+
+
+class MetaEstimatorBase(EstimatorBase):
+  """Defines the protocol for meta-estimation.
+
+  Meta-estimation uses a cardinality estimator in order to estimate bucket
+  indices in a Bloom Filter set to 1. If the cardinality estimator in question
+  is an exact set, this estimator is equivalent to the Bloom Filter base
+  estimator.
+  """
+
+  def __init__(self,
+               meta_sketch_factory,
+               meta_sketch_estimator,
+               adbf_estimator):
+    """Initializes a MetaEstimator using the given FirstMomentEstimator method.
+
+    Args:
+      meta_sketch_factory: a sketch factory method used to estimate Bloom Filter
+        buckets.
+      meta_sketch_estimator: an estimator to estimate the union cardinality
+        given a list of meta sketches.
+      adbf_estimator: an estimator to estimate the cardinality of a list of
+        AnyDistributionBloomFilter sketches.
+    """
+    super().__init__()
+    self.adbf_estimator = adbf_estimator
+    self.meta_sketch_factory = meta_sketch_factory
+    self.meta_sketch_estimator = meta_sketch_estimator
+
+  @classmethod
+  def _check_compatibility(cls, sketch_list):
+    """Determines if all sketches are compatible."""
+    for sketch in sketch_list:
+      assert isinstance(sketch, bf.AnyDistributionBloomFilter), (
+          'MetaEstimator must be used with an AnyDistributionBloomFilter.')
+
+    first_sketch = sketch_list[0]
+
+    for cur_sketch in sketch_list[1:]:
+      first_sketch.assert_compatible(cur_sketch)
+
+  def __call__(self, sketch_list):
+    """Transform AnyDistributionBloomFilters into meta sketch and estimate union."""
+    MetaEstimatorBase._check_compatibility(sketch_list)
+    if not sketch_list:
+      return 0
+
+    meta_sketches = self._transform_adbf_into_meta_sketches(sketch_list)
+    num_ones = self._estimate_num_ones(meta_sketches)
+    adbf_sketch = MetaEstimatorBase._construct_fake_adbf(
+        num_ones, sketch_list[0])
+    return self.adbf_estimator([adbf_sketch])
+
+  def _transform_adbf_into_meta_sketches(self, sketch_list):
+    """Transform AnyDistributionBloomFilters into meta sketch.
+
+    Args:
+      sketch_list: a list of AnyDistributionBloomFilter sketches.
+
+    Returns:
+      A list of meta sketches.
+    """
+    if not sketch_list:
+      return []
+
+    random_seed = sketch_list[0].random_seed
```

Comment:
It is possible, yet I doubt if it would be more useful. This implementation uses one random seed for all sketches. If we pass a random seed here, we are introducing another random seed. For reproducibility, the only difference is the one random seed vs two.

But arguably, the items inserted to the ADBF sketch and the meta sketch are different, the former being the raw IDs, and the latter being the indices of the active registers. So it would not introduce unknown correlation between the ADBF sketches and the meta sketches I think.

Please let me know if we need this change. If not, I would prefer this simple version.

Reply:
My concern is more around reaching into another class to extract this vs. having it passed in. In practice the values should be the same regardless I think. All that said, we can leave it as is.